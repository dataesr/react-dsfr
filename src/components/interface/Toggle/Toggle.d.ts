import * as React from 'react';

export type ToggleClassName = string | Object | any[];

export interface ToggleProps {
    onChange?: (...args: any[])=>any;
    id?: string;
    className?: ToggleClassName;
    hasSeparator?: boolean;
    hasLabelLeft?: boolean;
    description?: string;
    label: string;
    toggleColor?: string;
}

declare const Toggle: React.FC<ToggleProps>;

export default Toggle;

