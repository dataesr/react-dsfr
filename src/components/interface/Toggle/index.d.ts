import type Toggle from './Toggle';

export type { Toggle };
