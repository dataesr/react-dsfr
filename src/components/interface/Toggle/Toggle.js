import React, { useEffect, useRef } from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { v4 as uuidv4 } from 'uuid';
import dataAttributes from '../../../utils/data-attributes';

/*
* DSFR v1.3.1
*/
import '@gouvfr/dsfr/dist/component/toggle/toggle.css';

/**
 *
 * @visibleName Toggle
 */
const Toggle = ({
  className,
  hasSeparator,
  hasLabelLeft,
  label,
  id,
  toggleColor,
  onChange,
  description,
  ...remainingProps
}) => {
  const _id = useRef(id || uuidv4());
  const _className = classNames('fr-toggle', {
    'fr-toggle--border-bottom': hasSeparator,
    'fr-toggle--label-left': hasLabelLeft,
    'ds-fr-toggle': toggleColor && !remainingProps.disabled,
  }, className);

  useEffect(() => {
    _id.current = id || uuidv4();
  }, [id]);

  useEffect(() => {
    if (toggleColor) {
      document.documentElement.style.setProperty('--toggle-color', toggleColor);
    }
  }, [toggleColor]);

  return (
    <div
      className={_className}
      {...dataAttributes.getAll(remainingProps)}
    >
      <input
        onChange={onChange}
        type="checkbox"
        className="fr-toggle__input"
        id={_id.current}
      />
      <label
        className="fr-toggle__label"
        htmlFor={_id.current}
        data-fr-checked-label="Activé"
        data-fr-unchecked-label="Désactivé"
      >
        {label}
      </label>
      {description && <p className="fr-hint-text">{description}</p>}
    </div>
  );
};

Toggle.defaultProps = {
  id: '',
  toggleColor: '',
  className: '',
  hasSeparator: false,
  hasLabelLeft: false,
  description: '',
  onChange: undefined,
};

Toggle.propTypes = {
  onChange: PropTypes.func,
  id: PropTypes.string,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  hasSeparator: PropTypes.bool,
  hasLabelLeft: PropTypes.bool,
  description: PropTypes.string,
  label: PropTypes.string.isRequired,
  toggleColor: PropTypes.string,
};

export default Toggle;
