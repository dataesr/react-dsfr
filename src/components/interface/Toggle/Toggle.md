## Toggle - Interrupteur
L’usage des interrupteurs est à privilégier pour paramétrer des fonctionnalités transverses. Le changement d'état de l’interrupteur ne nécessite pas de validation.

### Toggle simple
```jsx
<Toggle
    label="Toggle Label"
/>
```
### Usage avec React state
```jsx
import { useState } from 'react';
const [checked, setCheched] = useState(false);

<Toggle
    checked={checked}
    onChange={() => setCheched(!checked)}
    label="Toggle Label"
/>
```

### Toggle avec séparateur et label à gauche
```jsx
<Toggle
    checked={false}
    label="Toggle Label"
    hasSeparator
    hasLabelLeft
/>
```

### Toggle disabled
```jsx
<Toggle
    disabled
    checked
    label="Toggle Label"
    description="Toggle descirption"
/>
```
