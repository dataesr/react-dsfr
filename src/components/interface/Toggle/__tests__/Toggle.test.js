import { render, screen, fireEvent } from '@testing-library/react';
import { v4 as uuidv4 } from 'uuid';
import Toggle from '..';

jest.mock('uuid', () => ({
  v4: jest.fn(),
}));

describe('<Toggle />', () => {
  beforeEach(() => {
    uuidv4.mockImplementationOnce(() => 'xxxxxxx');
  });
  it('renders Toggle properly', () => {
    const onChange = jest.fn();
    render(
      <Toggle
        isChecked
        onChange={onChange}
        label="Accepter les cookies"
        description="Activez-moi!"
        hasSeparator
        hasLabelLeft
      />,
    );
    const toggle = screen.getByTestId('toggle');
    expect(toggle.className).toBe('rf-toggle rf-toggle--border-bottom rf-toggle--label-left');
    expect(toggle).toBeInTheDocument();
    expect(toggle).toMatchSnapshot();
  });

  it('should fire onChange on click', () => {
    const onChange = jest.fn();
    render(
      <Toggle
        isChecked
        onChange={onChange}
        label="Accepter les cookies"
        description="Activez-moi!"
        hasSeparator
        hasLabelLeft
      />,
    );
    fireEvent.click(screen.getByTestId('toggle-input'));
    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
