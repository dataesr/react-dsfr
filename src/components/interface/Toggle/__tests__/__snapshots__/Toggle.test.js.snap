// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should render Toggle properly 1`] = `
<div
  className="fr-toggle fr-toggle--border-bottom fr-toggle--label-left"
  data-testid="toggle"
>
  <input
    className="fr-toggle__input"
    id="xyx"
    type="checkbox"
  />
  <label
    className="fr-toggle__label"
    data-fr-checked-label="Activé"
    data-fr-unchecked-label="Désactivé"
    htmlFor="xyx"
  >
    Accepter les cookies
  </label>
  <p
    className="fr-hint-text"
  >
    Activez-moi!
  </p>
</div>
`;
